### PR TITLE
Push multi-arch images

### LINKED ISSUE #162: Quick start not working for Mac M1 laptops
**Is your feature request related to a problem? If so, please describe.**

When trying the quick start guide, model mesh controller pod and model mesh serving pod are in pending state due to node affinity rules, it has only amd64

**Describe your proposed solution**
Adding the arm64 option in the nodeSelector values will be suffice i guess

**Describe alternatives you have considered**
I tried to edit the deployments and added the arm64 option

**Additional context**

Related issue #231

### LINKED ISSUE #231: 0/1 nodes are available: 1 node(s) didn't match Pod's node affinity/selector
I am setting up on MiniKube as well as KIND cluster. In both cases modelmesh-controller is Pending status: 

```
NAME                                    READY   STATUS    RESTARTS   AGE
etcd-c55788c9-g77js                     1/1     Running   0          2m17s
minio-646b7c5dd4-77kdm                  1/1     Running   0          2m17s
modelmesh-controller-5c5d84b956-hxgh9   0/1     Pending   0          114s
```

kubectl describe showing : 0/1 nodes are available: 1 node(s) didn't match Pod's node affinity/selector

Do you have any thought?

### PR DESCRIPTION
**Pull request by ddelange**
#### Motivation

Multi-platform build for `modelmesh-controller` and `modelmesh-controller-develop` images; enable minikube deployments on Mac M1 laptops (#162, #231)

#### Modifications

- remove `amd64` arch-specific `nodeAffinity` for modelmesh controller (#162, #231)
- build for platforms `linux/amd64`, `linux/arm64`, `linux/ppc64le`, `linux/s390x`
- use `ubi8/go-toolset:1.18` for developer image instead of `ubi8/ubi-minimal:8.7`
- build on `pull_request` as well as on `push` to catch build breaks during PR review
- push developer images to DockerHub on PR merge, use local registry for PR builds
- update deprecated `checkout` action
- fix infinite docker inside docker error when running `make run fmt` inside the developer container
- update `build_devimage.sh` script to only pull developer image if not already present
- update `develop.sh` script to use the `.develop_image_name`

#### Related PRs

- [x] [`modelmesh`](https://github.com/kserve/modelmesh/) (https://github.com/kserve/modelmesh/pull/80)
- [x] [`modelmesh-runtime-adapter`](https://github.com/kserve/modelmesh-runtime-adapter/) (https://github.com/kserve/modelmesh-runtime-adapter/pull/38)
- [x] [`rest-proxy`](https://github.com/kserve/rest-proxy/) (https://github.com/kserve/rest-proxy/pull/24)
- [x] [`modelmesh-minio-examples`](https://github.com/kserve/modelmesh-minio-examples/) (https://github.com/kserve/modelmesh-minio-examples/pull/3)

#### Related issues

Resolves #162
Resolves #231

### CODE REVIEW

File: .github/workflows/build-and-push.yml [hunk 1]
```diff
@@ -12,39 +12,68 @@ on:
...
         run: ./scripts/develop.sh make test
-      - name: Build runtime image
-        run: make build
-      - name: Log in to Docker Hub
-        run: docker login -u ${{ secrets.DOCKER_USER }} -p ${{ secrets.DOCKER_ACCESS_TOKEN }}
-      - name: Push to Docker Hub
+  build:
+    needs: test
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push'
```
Comment (ddelange) [inline .github/workflows/build-and-push.yml:33]: build not push on PRs https://github.com/kserve/rest-proxy/pull/22
```suggestion
```

Comment (njhill) [reply .github/workflows/build-and-push.yml]: @ckadner

**Comment by ckadner** [comment]
I think we need to widen the scope of this change to ...

- move to multi-stage docker build to instead of having two Dockerfiles (see current [build error](https://github.com/kserve/modelmesh-serving/actions/runs/3930877238/jobs/6809499875#step:7:137))
- use [ubi8/go-toolset](https://catalog.redhat.com/software/containers/ubi8/go-toolset/5ce8713aac3db925c03774d1?container-tabs=gti) as base image instead of `ubi8/minimal` to avoid complexities of building (or even just installing) go inside the Dockerfile

**Comment by ckadner** [comment]
Hi @ddelange, are you still planing to work on the remaining multi-arch PRs? We had another request to support multiple (non-amd64) platforms since. What do you think about me pushing some updates to your PRs to address some of the outstanding work -- if you don't mind me becoming a co-author on your PRs.

**Comment by ddelange** [comment]
Hi @ckadner feel free to take them over! I'm pretty side-tracked at work at the moment, but I'll be around, feel free to ping me!

**Comment by ddelange** [comment]
woop woop great work @ckadner !

**Comment by ckadner** [comment]
> woop woop great work @ckadner !

Sorry for hijacking your fork and branch for this @ddelange -- I wanted to make sure you remain the originator of this PR.

**Comment by ddelange** [comment]
no sorries and thank you! 🚀

File: .github/workflows/build-and-push.yml [hunk 2]
```diff
@@ -4,47 +4,150 @@ on:
...
+
+          DEV_IMAGE="docker.io/${{ env.DEV_IMAGE_NAME }}:$DEV_IMG_TAG"            
+          DEV_IMAGE_EXISTS=false
+
+          # Check if the developer image already exists in the registry. The
+          # developer image doesn't change often and is costly to build. We want
+          # to avoid rebuilding it unnecessarily. If we need to build it and if
+          # this workflow is running on a forked PR, then we must use the local
+          # image registry.
+          if $(docker manifest inspect $DEV_IMAGE >/dev/null ); then
```
Comment (ckadner) [inline .github/workflows/build-and-push.yml:98]:
```bash
$ docker buildx imagetools inspect kserve/modelmesh-controller-develop:e0ec890deaa45961 --format "{{json .Manifest}}" | grep architecture
 | grep "amd64\|arm64\|ppc64le\|s390x"
```
```JSON
        "architecture": "amd64",
        "architecture": "arm64",
        "architecture": "ppc64le",
        "architecture": "s390x",
```

Comment (ckadner) [reply .github/workflows/build-and-push.yml:98]: yay, it found it:

```
> Run # ...
  
  if $(docker manifest inspect $DEV_IMAGE >/dev/null ); then
    DEV_IMAGE_EXISTS=true
  else
    if [ "pull_request" == "pull_request" ]; then
      DEV_IMAGE="localhost:5000/kserve/modelmesh-controller-develop:$DEV_IMG_TAG"
    fi
  fi
  
  ...
  # print env vars for debugging
  ...
DEV_IMAGE=docker.io/kserve/modelmesh-controller-develop:e0ec890deaa45961
DEV_IMAGE_EXISTS=true
```

**Comment by ddelange** [comment]
I think something is wrong with the [`dockersize`](https://stackoverflow.com/a/73108928/5511061) 🤔 
```console
$ dockersize kserve/modelmesh-controller-develop:e0ec890deaa45961
linux/amd64    928.67M
linux/arm64    902.75M
linux/ppc64le  924.50M
linux/s390x    908.23M
$ dockersize kserve/modelmesh-controller
linux/amd64  60.42M
```

**Comment by ckadner** [comment]
> I think something is wrong with the [`dockersize`](https://stackoverflow.com/a/73108928/5511061) 🤔
> 
> ```
> $ dockersize kserve/modelmesh-controller-develop:e0ec890deaa45961
> linux/amd64    928.67M
> linux/arm64    902.75M
> linux/ppc64le  924.50M
> linux/s390x    908.23M
> $ dockersize kserve/modelmesh-controller
> linux/amd64  60.42M
> ```

Looks okay on DockerHub, change in size is minimal:
image

**Comment by ckadner** [comment]
The small optimization to check if the developer image exists (instead of always building it), shaves about 30 mins off the overall job completion time, but the [arch-specific `go` build times](https://github.com/kserve/modelmesh-serving/actions/runs/4560241237/jobs/8045002449#step:12:492) for the runtime image take a horrendous 50 mins:

```
    #37 [linux/amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
    #37 DONE 349.6s

    #43 [linux/ppc64le->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
    #43 DONE 2622.6s

    #44 [linux/arm64->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
    #44 DONE 2793.7s

    #26 [linux/s390x->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
    #26 DONE 3020.3s
```

I need to read up on how to [optimize](https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/) that.

**Comment by ckadner** [comment]
> The optimization to [check if the developer image exists](https://github.com/kserve/modelmesh-serving/blob/9bb9586560ba3a63426fa016dadf9a0a74890d99/.github/workflows/build-and-push.yml#L98-L104) (instead of always building it), shaves about 30 mins off the overall job completion time, but the [arch-specific `go` build](https://github.com/kserve/modelmesh-serving/actions/runs/4560241237/jobs/8045002449#step:12:492) for the controller image take a horrendous 50 mins:
> 
> ```
>     #37 [linux/amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
>     #37 DONE 349.6s
> 
>     #43 [linux/ppc64le->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
>     #43 DONE 2622.6s
> 
>     #44 [linux/arm64->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
>     #44 DONE 2793.7s
> 
>     #26 [linux/s390x->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
>     #26 DONE 3020.3s
> ```
> 
> I need to read up on how to [optimize](https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/) that.

Woohoo, the multi-platform build times are down to [13.5 minutes](https://github.com/kserve/modelmesh-serving/actions/runs/4602956469), or [8.5 minutes if the developer image exists](https://github.com/kserve/modelmesh-serving/actions/runs/4602842301).

**Comment by ckadner** [comment]
> > The optimization to [check if the developer image exists](https://github.com/kserve/modelmesh-serving/blob/9bb9586560ba3a63426fa016dadf9a0a74890d99/.github/workflows/build-and-push.yml#L98-L104) (instead of always building it), shaves about 30 mins off the overall job completion time, but the [arch-specific `go` build](https://github.com/kserve/modelmesh-serving/actions/runs/4560241237/jobs/8045002449#step:12:492) for the controller image take a horrendous 50 mins:
> > ```
> >     #37 [linux/amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
> >     #37 DONE 349.6s
> > 
> >     #43 [linux/ppc64le->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
> >     #43 DONE 2622.6s
> > 
> >     #44 [linux/arm64->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
> >     #44 DONE 2793.7s
> > 
> >     #26 [linux/s390x->amd64 build 7/7] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
> >     #26 DONE 3020.3s
> > ```
> > 
> > 
> >   
> > I need to read up on how to [optimize](https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/) that.
> 
> Woohoo, the multi-platform build times are down to [13.5 minutes](https://github.com/kserve/modelmesh-serving/actions/runs/4602956469), or [8.5 minutes if the developer image exists](https://github.com/kserve/modelmesh-serving/actions/runs/4602842301).

Thanks @tjohnson31415 for pointing out the error with setting "default" `ARG`s for `TARGETOS` and `TARGETARCH` in the Dockerfile which had the effect of building `amd64` binaries for all platforms.

Updated, now we actually build target platform specific binaries (with comparable cross-compile build times):

```
#22 [linux/amd64->s390x build 7/7] RUN --mount=type=cache,target=/root/.cache/go-build     --mount=type=cache,target=/go/pkg     CGO_ENABLED=0 GOOS=linux GOARCH=s390x GO111MODULE=on go build -a -o manager main.go
#22 DONE 421.5s

#23 [linux/amd64->arm64 build 7/7] RUN --mount=type=cache,target=/root/.cache/go-build     --mount=type=cache,target=/go/pkg     CGO_ENABLED=0 GOOS=linux GOARCH=arm64 GO111MODULE=on go build -a -o manager main.go
#23 DONE 423.5s

#24 [linux/amd64 build 7/7] RUN --mount=type=cache,target=/root/.cache/go-build     --mount=type=cache,target=/go/pkg     CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
#24 DONE 426.9s

#21 [linux/amd64->ppc64le build 7/7] RUN --mount=type=cache,target=/root/.cache/go-build     --mount=type=cache,target=/go/pkg     CGO_ENABLED=0 GOOS=linux GOARCH=ppc64le GO111MODULE=on go build -a -o manager main.go
#21 DONE 426.9s
```

**Review by tjohnson31415 — APPROVED** [review]
LGTM

**Pull request merged**